File: 6.5.0/src/c++/common/syncml/core/ResponseCommand.cpp
/*
 * Copyright (C) 2003-2007 Funambol, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY, TITLE, NONINFRINGEMENT or FITNESS FOR A PARTICULAR
 * PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
 * 02111-1307  USA
 */


#include "syncml/core/ResponseCommand.h"

ResponseCommand::ResponseCommand() {
    msgRef     = NULL;
    cmdRef     = NULL;
    targetRef  = new ArrayList(); // TargetRef[]
    sourceRef  = new ArrayList(); // SourceRef[]

}
ResponseCommand::~ResponseCommand()  {
    if (msgRef)     { delete [] msgRef;     msgRef    = NULL; }
    if (cmdRef)     { delete [] cmdRef;     cmdRef    = NULL; }
    if (targetRef)  { targetRef->clear();   } //delete targetRef; targetRef = NULL; }
    if (sourceRef)  { sourceRef->clear();   } //delete sourceRef; sourceRef = NULL; }

}

/**
 * Creates a new ResponseCommand object.
 *
 * @param cmdID the command idendifier  - NOT NULL
 * @param msgRef message reference
 * @param cmdRef command reference - NOT NULL
 * @param targetRefs target references
 * @param sourceRefs source references
 * @param items command items
 * if any of the NOT NULL parameter is null
 */
ResponseCommand::ResponseCommand(CmdID*         cmdID     ,
                                 const char*   msgRef    ,
                                 const char*   cmdRef    ,
                                 ArrayList*     targetRefs,
                                 ArrayList*     sourceRefs,
                                 ArrayList*     items      )
: ItemizedCommand (cmdID, items)
{
    this->msgRef     = stringdup(msgRef);
    this->cmdRef     = stringdup(cmdRef);
    this->targetRef  = targetRefs->clone();
    this->sourceRef  = sourceRefs->clone();
}


/**
 * Returns the message reference
 *
 * @return the message reference
 *
 */
const char* ResponseCommand::getMsgRef() {
    return msgRef;
}

/**
 * Sets the message reference
 *
 * @param msgRef message reference
 */
void ResponseCommand::setMsgRef(const char* msgRef) {
     if (this->msgRef) {
        delete [] this->msgRef; this->msgRef = NULL;
    }
    this->msgRef = stringdup(msgRef);
}

/**
 * Returns the command reference
 *
 * @return the command reference
 *
 */
const char* ResponseCommand::getCmdRef() {
    return cmdRef;
}

/**
 * Sets the command reference
 *
 * @param cmdRef commandreference - NOT NULL
 *
 */
void ResponseCommand::setCmdRef(const char* cmdRef) {
     if (this->cmdRef) {
         delete [] this->cmdRef; this->cmdRef = NULL;
     }
     this->cmdRef = stringdup(cmdRef);
}

/**
 * Returns the target references
 *
 * @return the target references
 *
 */
ArrayList* ResponseCommand::getTargetRef() {
    return targetRef;
}

/**
 * Sets the target references
 *
 * @param targetRefs target refrences
 */
void ResponseCommand::setTargetRef(ArrayList* targetRefs) {
    if (this->targetRef) {
        this->targetRef->clear();
    }
    if (targetRefs != NULL) {
        this->targetRef = targetRefs->clone();
    }
}

/**
 * Returns the source references
 *
 * @return the source references
 *
 */
ArrayList* ResponseCommand::getSourceRef(){
    return sourceRef;
}

/**
 * Sets the source references
 *
 * @param sourceRefs source refrences
 */
void ResponseCommand::setSourceRef(ArrayList* sourceRefs) {
    if (this->sourceRef) {
        this->sourceRef->clear();
    }
    if (sourceRefs != NULL) {
        this->sourceRef = sourceRefs->clone();
    }
}

File: us8294/syncml/src/main/java/com/funambol/syncml/protocol/ResponseCommand.java
/*
 * Funambol is a mobile platform developed by Funambol, Inc.
 * Copyright (C) 2003 - 2007 Funambol, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License version 3 as published by
 * the Free Software Foundation with the addition of the following permission
 * added to Section 15 as permitted in Section 7(a): FOR ANY PART OF THE COVERED
 * WORK IN WHICH THE COPYRIGHT IS OWNED BY FUNAMBOL, FUNAMBOL DISCLAIMS THE
 * WARRANTY OF NON INFRINGEMENT  OF THIRD PARTY RIGHTS.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program; if not, see http://www.gnu.org/licenses or write to
 * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301 USA.
 *
 * You can contact Funambol, Inc. headquarters at 643 Bair Island Road, Suite
 * 305, Redwood City, CA 94063, USA, or at email address info@funambol.com.
 *
 * The interactive user interfaces in modified source and object code versions
 * of this program must display Appropriate Legal Notices, as required under
 * Section 5 of the GNU Affero General Public License version 3.
 *
 * In accordance with Section 7(b) of the GNU Affero General Public License
 * version 3, these Appropriate Legal Notices must retain the display of the
 * "Powered by Funambol" logo. If the display of the logo is not reasonably
 * feasible for technical reasons, the Appropriate Legal Notices must display
 * the words "Powered by Funambol".
 */


package com.funambol.syncml.protocol;

import java.util.*;

/**
 * This is the base (and abstract) class for response commands
 */
public abstract class ResponseCommand extends ItemizedCommand {

    // ---------------------------------------------------------- Protected data

    /**
     * Message reference
     */
    protected String msgRef;

    /**
     * Command reference
     */
    protected String cmdRef;

    /**
     * Target references
     */
    protected Vector targetRef = new Vector();

    /**
     * Source references
     */
    protected Vector sourceRef = new Vector();

    // ------------------------------------------------------------ Constructors

    /**
     * For serialization purposes
     */
    protected ResponseCommand() {}

    /**
     * Creates a new ResponseCommand object.
     *
     * @param cmdID the command idendifier  - NOT NULL
     * @param msgRef message reference
     * @param cmdRef command reference - NOT NULL
     * @param targetRefs target references
     * @param sourceRefs source references
     * @param items command items
     *
     * @throws IllegalArgumentException if any of the NOT NULL parameter is null
     */
    public ResponseCommand(
        final String cmdID     ,
        final String            msgRef    ,
        final String            cmdRef    ,
        final Vector            targetRefs,
        final Vector            sourceRefs,
        final Vector            items     ) {
        super(cmdID, items);

        setCmdRef(cmdRef);

        this.msgRef = msgRef;

        setTargetRef(targetRefs);
        setSourceRef(sourceRefs);
    }

    // ---------------------------------------------------------- Public methods

    public void init() {
        super.init();

        msgRef = null;
        cmdRef = null;
        targetRef.removeAllElements();
        sourceRef.removeAllElements();
    }

    /**
     * Returns the message reference
     *
     * @return the message reference
     *
     */
    public String getMsgRef() {
        return this.msgRef;
    }

    /**
     * Sets the message reference
     *
     * @param msgRef message reference
     */
    public void setMsgRef(String msgRef) {
        this.msgRef = msgRef;
    }

    /**
     * Returns the command reference
     *
     * @return the command reference
     *
     */
    public String getCmdRef() {
        return cmdRef;
    }

    /**
     * Sets the command reference
     *
     * @param cmdRef commandreference - NOT NULL
     *
     * @throws IllegalArgumentException if cmdRef is null
     */
    public void setCmdRef(String cmdRef) {
        if (cmdRef == null) {
            throw new IllegalArgumentException("cmdRef cannot be null");
        }
        this.cmdRef = cmdRef;
    }

    /**
     * Returns the target references
     *
     * @return the target references
     *
     */
    public Vector getTargetRef() {
        return this.targetRef;
    }

    /**
     * Sets the target references
     *
     * @param targetRefs target refrences
     */
    public void setTargetRef(Vector targetRef) {
        this.targetRef = targetRef;
    }

    /**
     * Sets a single target reference (remove all existing ones)
     */
    public void setTargetRef(TargetRef ref) {
        Vector refs = new Vector();
        refs.addElement(ref);
        setTargetRef(refs);
    }

    /**
     * Returns the source references
     *
     * @return the source references
     *
     */
    public Vector getSourceRef() {
        return this.sourceRef;
    }

    /**
     * Sets the source references
     *
     * @param sourceRefs source refrences
     */
    public void setSourceRef(Vector sourceRef) {
        this.sourceRef = sourceRef;
    }

    /**
     * Sets a single target reference (remove all existing ones)
     */
    public void setSourceRef(SourceRef ref) {
        Vector refs = new Vector();
        refs.addElement(ref);
        setSourceRef(refs);
    }


    /**
     * Returns the command name. It must be redefined by subclasses.
     *
     * @return the command name
     */
    abstract public String getName();
}
